include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -1673,117 +1549,91 @@ struct __parallel_sort_submitter<__internal::__optional_kernel_name<_LeafSortNam
-            //  Legend:
-            //  * [] - border between pairs of sorted sequences which are to be merged
-            //  * @  - border between each sorted sequence in a pair
-            //  * || - border between chunks
-
-            _Size __sorted_pair = 2 * __sorted;
-            _Size __chunks_in_sorted = __sorted / __chunk;
-            _Size __full_pairs = __n / __sorted_pair;
-            _Size __incomplete_pair = __n - __sorted_pair * __full_pairs;
-            _Size __first_block_in_incomplete_pair = __incomplete_pair > __sorted ? __sorted : __incomplete_pair;
-            _Size __incomplete_last_chunk = __first_block_in_incomplete_pair % __chunk != 0;
-            _Size __incomplete_pair_steps = __first_block_in_incomplete_pair / __chunk + __incomplete_last_chunk;
-            _Size __full_pairs_steps = __full_pairs * __chunks_in_sorted;
-            _Size __steps = __full_pairs_steps + __incomplete_pair_steps;
-
-            __event1 = __exec.queue().submit([&, __data_in_temp, __sorted, __sorted_pair, __chunk, __chunks_in_sorted,
-                                              __steps](sycl::handler& __cgh) {
+            __event1 = __exec.queue().submit([&, __data_in_temp, __n, __comp](sycl::handler& __cgh) {
                 __cgh.depends_on(__event1);
+
                 oneapi::dpl::__ranges::__require_access(__cgh, __rng);
-                auto __temp_acc = __temp.template get_access<__par_backend_hetero::access_mode::read_write>(__cgh);
-                __cgh.parallel_for<_GlobalSortName...>(
-                    sycl::range</*dim=*/1>(__steps), [=](sycl::item</*dim=*/1> __item_id) {
-                        const _Size __idx = __item_id.get_linear_id();
-                        // Borders of the first and the second sorted sequences
-                        const _Size __start_1 = sycl::min(__sorted_pair * ((__idx * __chunk) / __sorted), __n);
-                        const _Size __end_1 = sycl::min(__start_1 + __sorted, __n);
-                        const _Size __start_2 = __end_1;
-                        const _Size __end_2 = sycl::min(__start_2 + __sorted, __n);
+                sycl::accessor __dst(__temp, __cgh, sycl::read_write, sycl::no_init);
 
-                        // Distance between the beginning of a sorted sequence and the beginning of a chunk
-                        const _Size __offset = __chunk * (__idx % __chunks_in_sorted);
+                __cgh.parallel_for<_GlobalSortName...>(sycl::range</*dim=*/1>(__steps), [=](sycl::item</*dim=*/1> __item_id)
+                    {
+                        const ::std::uint32_t i_elem = __item_id.get_linear_id() * __chunk;
+                        const auto i_elem_local = i_elem % (__n_sorted*2);
 
-                        if (!__data_in_temp)
+                        const auto offset = ::std::min<::std::uint32_t>((i_elem / (__n_sorted * 2)) * (__n_sorted * 2), __n);
```

Comment:
Actually... right about this we are already using `% (__n_sorted * 2)`.  We can just subtract `__i_elem_local`.

It could be faster to  use the function mentioned above to round down to multiple (power of 2), and then calculate `__i_elem_local` by subtracting out the rounded down value instead of using `%`, but that would have to be examined.

Reply:
https://github.com/oneapi-src/oneDPL/issues/1121 resolving this